support-frontend/assets/helpers/tracking/quantumMetric.ts

```diff
@@ -128,6 +138,68 @@ function sendEventWhenReadyTrigger(sendEventWhenReady: () => void): void {
 	}
 }
 
+function sendEventAcquisitionDataFromQueryParamEvent(
+	acquisitionData: ReferrerAcquisitionData,
+): void {
+	void ifQmPermitted(() => {
+		const sendEventWhenReady = () => {
+			const source = acquisitionData.source;
+			const componentId = acquisitionData.componentId;
+			const componentType = acquisitionData.componentType;
+			const campaignCode = acquisitionData.campaignCode;
+			const referrerUrl = acquisitionData.referrerUrl;
+			const isRemote = acquisitionData.isRemote;
+
+			if (window.QuantumMetricAPI?.isOn()) {
+				if (source) {
```

Comment:
There's some repetition here, for each key in acquisitionData that we want to log we're:

1) Checking the existence of a key/value in `acquisitionData`
2) Converting the value from `acquisitionData` to a string
3) Calling `sendEvent`

We could remove this by iterating over an object (eg. `acquisitionDataKeysToLog` below) and executing this repeated code inside a code block (eg. inside the `forEach`). The example below illustrates how this could work...

```
const acquisitionDataKeysToLog = {
  source: SendEventAcquisitionDataFromQueryParam.Source,
  componentId: SendEventAcquisitionDataFromQueryParam.ComponentId,
  componentType: SendEventAcquisitionDataFromQueryParam.ComponentType,
  campaignCode: SendEventAcquisitionDataFromQueryParam.CampaignCode,
  referrerUrl: SendEventAcquisitionDataFromQueryParam.ReferrerUrl,
  isRemote: SendEventAcquisitionDataFromQueryParam.IsRemote
}

Object.keys(acquisitionDataKeysToLog).forEach((key) => {
  const acquisitionDataValueToLog = acquisitionData[key]?.toString();

  if (acquisitionDataValueToLog) {
    sendEvent(
      acquisitionDataKeysToLog[key],
      false,
      acquisitionDataValueToLog
    );
  }
});
```

Reply:
I struggled with this,  kept getting `Unsafe assignment of an `any` value.` error on the key field ( type string not a keyof ReferrerAcquisitionData perhaps?) Lakshmi if you would like to pair keen to lok at this?

Reply:
I think this uses George's suggestion of declaring the `ReferrerAcquisitionDataKeys` type and the `keyof` operator.

```
type ReferrerAcquisitionDataKeys = {
	source: number;
	componentId: number;
	componentType: number;
	campaignCode: number;
	referrerUrl: number;
	isRemote: number;
};

function sendEventAcquisitionDataFromQueryParamEvent(
	acquisitionData: ReferrerAcquisitionData,
): void {
	void ifQmPermitted(() => {
		const sendEventWhenReady = () => {
			const source = acquisitionData.source;
			if (window.QuantumMetricAPI?.isOn()) {
				if (source) {
					const acquisitionDataKeysToLog: ReferrerAcquisitionDataKeys = {
						source: SendEventAcquisitionDataFromQueryParam.Source,
						componentId: SendEventAcquisitionDataFromQueryParam.ComponentId,
						componentType: SendEventAcquisitionDataFromQueryParam.ComponentType,
						campaignCode: SendEventAcquisitionDataFromQueryParam.CampaignCode,
						referrerUrl: SendEventAcquisitionDataFromQueryParam.ReferrerUrl,
						isRemote: SendEventAcquisitionDataFromQueryParam.IsRemote,
					};
					Object.keys(acquisitionDataKeysToLog).forEach((key) => {
						const acquisitionDataValueToLog =
							acquisitionData[key as keyof ReferrerAcquisitionData]?.toString();
						if (acquisitionDataValueToLog) {
							sendEvent(
								acquisitionDataKeysToLog[
									key as keyof ReferrerAcquisitionDataKeys
								],
								false,
								acquisitionDataValueToLog,
							);
						}
					});
				}
			}
		};

		sendEventWhenReadyTrigger(sendEventWhenReady);
	});
}
```

Reply:
Thanks, It would be simple to use instead of  type ReferrerAcquisitionDataKeys = Record<string, number>;    type ReferrerAcquisitionDataKeys = {
	source: number;
	componentId: number;
	componentType: number;
	campaignCode: number;
	referrerUrl: number;
	isRemote: number;
};